### PR TITLE
Unloading assignment bug

### PR DESCRIPTION
**Pull request by DomHenderson**
Previously, we handled compound assignments by converting x += 1 into x = x + 1 and leaving it at that. However this meant that we converted a[i++] += 1 into a[i++] = a[i++] + 1, which is not equivalent. This fixes this by extracting volatile subexpressions like i++, but still cloning things like index accesses. This is necessary because just extracting the lhs to a constant and cloning that would assign to the 'constant', not to the element in the array

### CODE REVIEW

File: src/passes/unloadingAssignment.ts [hunk 1]
```diff
@@ -63,4 +82,48 @@ export class UnloadingAssignment extends ASTMapper {
...
+        `Unexpected empty index access in compound assignment ${printNode(node)}`,
+      );
+      if (!(node.vIndexExpression instanceof Literal)) {
+        const decl = this.extractIndividualSubExpression(node.vIndexExpression, ast);
+        return [decl, ...this.extractUnstableSubexpressions(node.vBaseExpression, ast)];
+      }
+    } else if (node instanceof MemberAccess) {
+      return this.extractUnstableSubexpressions(node.vExpression, ast);
+    } else if (node instanceof FunctionCall) {
+      return [this.extractIndividualSubExpression(node, ast)];
```
Comment (GuyMcComb) [inline src/passes/unloadingAssignment.ts:109]: Would be interested to see an example of Function(i++) being used as an lvalue or am I missing what is going on here?

**Review by GuyMcComb — CHANGES_REQUESTED** [review]

File: src/passes/unloadingAssignment.ts [hunk 2]
```diff
@@ -63,4 +82,48 @@ export class UnloadingAssignment extends ASTMapper {
...
+  // node is the lvalue of a compound assignment
+  // this function extracts subexpressions that are not guaranteed to produce the same result every time
+  // to constants to evaluate once before the assignment
+  // this is needed because extracting to a 'constant' and then assigning to it would not modify the correct variable
+  // we also extract non-trivial expressions so they're only calculated once, though this is not technically necessary
+  // an array of created statements are returned so they can be visited
+  private extractUnstableSubexpressions(
+    node: Expression,
+    ast: AST,
+  ): VariableDeclarationStatement[] {
```
[the remark this answers is not in the document]
Comment (DomHenderson) [reply src/passes/unloadingAssignment.ts:95]: The return value is used on line 29 when we iterate through the returned array and dispatchVisit each one. Without this you can extract increment operators from inside compound assignments and then not process them

File: src/passes/unloadingAssignment.ts (hunk 1 above)
Comment (GuyMcComb) [reply src/passes/unloadingAssignment.ts:109]: Oh I tough you were saying `getArray(i++)`

**Review by GuyMcComb — APPROVED** [review]

**Comment by DomHenderson** [comment]
Converted to draft whilst I check one final case

**Comment by DomHenderson** [comment]
Case I was concerned about is not actually valid (assignments are not lvalues), so nothing more to be added since the review

**Pull request merged**